src/Compilers/CSharp/Portable/Parser/Lexer.cs

```diff
@@ -2293,12 +2293,144 @@ private void LexSyntaxTrivia(bool afterFirstToken, bool isTrailing, ref SyntaxLi
+
+        private bool IsConflictMarkerTrivia()
+        {
+            var position = TextWindow.Position;
+            var text = TextWindow.Text;
+            if (position == 0 || SyntaxFacts.IsNewLine(text[position - 1]))
+            {
+                var firstCh = text[position];
+
+                if ((position + s_conflictMarkerLength) <= text.Length)
+                {
+                    for (int i = 0, n = s_conflictMarkerLength; i < n; i++)
+                    {
+                        if (text[position + i] != firstCh)
+                        {
+                            return false;
+                        }
+                    }
+
+                    if (firstCh == '=')
+                    {
+                        return true;
+                    }
+
+                    return (position + s_conflictMarkerLength) < text.Length &&
+                        text[position + s_conflictMarkerLength] == ' ';
+                }
+            }
+
+            return false;
+        }
+
+        private void LexConflictMarkerTrivia(ref SyntaxListBuilder triviaList)
+        {
+            this.AddError(TextWindow.Position, s_conflictMarkerLength,
+                ErrorCode.ERR_Merge_conflict_marker_encountered);
+
+            var startCh = this.TextWindow.PeekChar();
+
+            // First create a trivia from the start of this merge conflict marker to the
+            // end of line/file (whicever comes first).
+            LexConflictMarkerHeader(ref triviaList);
+
+            // Now add the newlines as the next trivia.
+            LexConflictMarkerEndOfLine(ref triviaList);
+
+            // Now, if it was an ======= marker, then also created a DisabledText trivia for
+            // the contents of the file after it, up until the next >>>>>>> marker we see.
+            if (startCh == '=')
+            {
+                LexConflictMarkerDisabledText(ref triviaList);
+            }
+        }
+
+        private SyntaxListBuilder LexConflictMarkerDisabledText(ref SyntaxListBuilder triviaList)
+        {
+            // Consume everything from the start of the mid-conflict marker to the start of the next
+            // end-conflict marker.
+            this.Start();
+
+            while (true)
+            {
+                var ch = this.TextWindow.PeekChar();
+                if (ch == SlidingTextWindow.InvalidCharacter)
+                {
+                    break;
+                }
+
+                if (ch == '>' && IsConflictMarkerTrivia())
+                {
+                    break;
+                }
+
+                this.TextWindow.AdvanceChar();
+            }
+
+            if (this.TextWindow.Width > 0)
+            {
+                this.AddTrivia(SyntaxFactory.DisabledText(TextWindow.GetText(false)), ref triviaList);
+            }
+
+            return triviaList;
+        }
+
+        private void LexConflictMarkerEndOfLine(ref SyntaxListBuilder triviaList)
```

Comment:
Maybe name this `LexConflictMarkerEndLines`?
But I don't understand why more than one newline after `<<<<<<<` would be trailing trivia on it. Any subsequent newlines feel like they *belong* to the first token in that half of the diff (so that a codefix taking everything in the "mine" side wouldn't also need to worry about extra newlines attached to the `<<<<<<<` line -- they could just take everything in that half of the diff and know that it's correct.